File: AutomataConverter/src/main/kotlin/DeterministicFiniteAutomata.kt
package AutomataConverter

import java.util.*

private data class StateSet(val states: Collection<State>){
    val isAcceptingState: Boolean = states.any {it.isAcceptingState}

    override fun equals(other: Any?): Boolean {
        return if (other is StateSet) states.size == other.states.size && states.containsAll(other.states) else false
    }

    override fun hashCode(): Int {
        return states.hashCode()
    }
}

private data class IntermediateTransition(val from: StateSet, val via: Char, val to: StateSet)

class DeterministicFiniteAutomata(cardinality: Int,
                        validTokens: String,
                        acceptingStates: Array<Int>,
                        startState: Int,
                        transitionMap: List<Transition>
) : Automata(cardinality, validTokens, acceptingStates, startState, transitionMap) {
    companion object Converter{
        fun fromNFA(nfa: NonFiniteAutomata): DeterministicFiniteAutomata {
            val newStartState = StateSet(listOf(State(nfa.startState, isStartState = true)))
            val visitedSets = ArrayList<StateSet>()
            val toInspect = Stack<StateSet>()
            toInspect.push(newStartState)

            val intermediateTransitions = ArrayList<IntermediateTransition>()

            do
            {
                val set = toInspect.pop()
                visitedSets.add(set)

                for(c in nfa.validTokens)
                {
                    val reachableStates = ArrayList<State>()
                    // Construct a state set of all reachable states from all states in the current set via c
                    for(s in set.states)
                    {
                        nfa.transitionMap.filter { it.from == s && it.via == c }.forEach { reachableStates.add(it.to) }
                    }

                    // If we have not visited this state set yet push it onto the stack
                    val toState = StateSet(reachableStates.distinctBy { it.id })
                    if(!visitedSets.contains(toState)) toInspect.push(toState)

                    intermediateTransitions.add(IntermediateTransition(set, c, toState))
                }
            }while(toInspect.isNotEmpty())

            // Map each visited set of states to new IDs and build the new state set and transition map
            return DeterministicFiniteAutomata(
                    visitedSets.size,
                    nfa.validTokens,
                    visitedSets.mapIndexed { i, stateSet -> if(stateSet.isAcceptingState) i else -1 }.filter { it > 0 }.toTypedArray(),
                    0,
                    intermediateTransitions.map {
                        Transition(
                                State(
                                        visitedSets.indexOf(it.from)
                                ),
                                it.via,
                                State(
                                        visitedSets.indexOf(it.to)
                                )
                        )
                    }
            )
        }
    }
}

File: src/AutomataConverter/DeterministicFiniteAutomata.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AutomataConverter
{
    /// <summary>
    /// A Deterministic, Finite automata
    /// </summary>
    public class DeterministicFiniteAutomata : Automata
    {
        public DeterministicFiniteAutomata(int cardinality, string tokens, IEnumerable<int> acceptingStates, int startState, Dictionary<int, IEnumerable<Transition>> transitionMap)
            : base(cardinality, tokens, acceptingStates, startState, transitionMap)
        {
            // Ensure that the transition map that was provided is deterministic
            foreach(var state in TransitionMap.Keys)
            {
                if(TransitionMap[state].Count() != ValidTokens.Length)
                    throw new ArgumentException(nameof(transitionMap), $"The transition map does not contain the correct number of transitions for state {state}. It should have exactly {ValidTokens.Length} transitions from it, otherwise it is not deterministic");
            }
        }
    }

    public static class NonFiniteAutomataExtensions
    {
        /// <summary>
        /// A collection of states. Basically a list where <code>.Equals</code> checks
        /// that the two lists contain the same elements
        /// </summary>
        private class StateSet : List<int>
        {
            public StateSet(IEnumerable<int> states) : base(states)
            {
            }

            public override bool Equals(object other)
            {
                if(other is StateSet)
                {
                    return ((StateSet)other).OrderBy(x => x).SequenceEqual(this.OrderBy(x => x));
                }
                else
                {
                    return false;
                }
            }
        }

        /// <summary>
        /// An intermediate transition used when building the DFA
        /// </summary>
        private class IntermediateTransition
        {
            /// <summary>
            /// The set of states the transition starts from
            /// </summary>
            public readonly StateSet From;

            /// <summary>
            /// The token that activates the transition
            /// </summary>
            public readonly char Via;

            /// <summary>
            /// The set of states the transition maps to
            /// </summary>
            public readonly StateSet To;

            internal IntermediateTransition(StateSet from, char via, StateSet to)
            {
                From = from;
                Via = via;
                To = to;
            }
        }

        /// <summary>
        /// Converts the NFA to DFA using repeated subset construction
        /// </summary>
        /// <param name="nfa">the NFA to convert</param>
        /// <returns>A deterministic, finite automata</return>
        public static DeterministicFiniteAutomata convertToDFA(this NonDeterministicFiniteAutomata nfa)
        {
            var newStartState = new StateSet(new List<int>{nfa.StartState});
            var visitedSets = new List<StateSet>();
            var toInspect = new Stack<StateSet>();

            // Start at the start state
            toInspect.Push(newStartState);

            var intermediateTransitions = new List<IntermediateTransition>();

            do // while there are more states to inspect
            {
                // Mark this set as visited
                var set = toInspect.Pop();
                visitedSets.Add(set);

                foreach(var c in nfa.ValidTokens)
                {
                    var reachableStates = new List<int>();

                    // Construct a state set of all reachable states from all states in the current set via c
                    foreach(var s in set)
                    {
                        if(!nfa.TransitionMap.ContainsKey(s)) continue;
                        reachableStates.AddRange(nfa.TransitionMap[s].Where(t => t.Via == c).Select(t => t.To));
                    }

                    // If we have not visited this state set yet push it onto the stack
                    var toState = new StateSet(reachableStates.Distinct());
                    if(!visitedSets.Contains(toState)) toInspect.Push(toState);

                    // Map the transition
                    intermediateTransitions.Add(new IntermediateTransition(set, c, toState));
                }
            }while(toInspect.Count > 0);

            // Map each visited set of states to new IDs and build the new state set and transition map
            return new DeterministicFiniteAutomata(
                visitedSets.Count,
                nfa.ValidTokens,
                visitedSets.Select((s, i) => nfa.AcceptingStates.Intersect(s).Any() ? i : -1).Where(i => i > 0),
                0,
                intermediateTransitions.Select(t => new Transition(
                    visitedSets.IndexOf(t.From),
                    t.Via,
                    visitedSets.IndexOf(t.To)
                )).GroupBy(t => t.From).ToDictionary(k => k.Key, v => v.AsEnumerable())
            );
        }
    }
}